### PR TITLE
Add Feature: state restoration

### PR DESCRIPTION
**Pull request by penguingovernor**
This commit also refactors the state package.
It adds an index.js file so that other modules can include
its contents by importing directly from the path.

### CODE REVIEW

File: src/state/REPLState.js [hunk 1]
```diff
@@ -94,13 +95,54 @@ class REPLState {
   static Decode(encodedState) {
     let jsonState = JSON.parse(encodedState);
     return new REPLState(
-      decodeBase64(jsonState.source),
-      decodeBase64(jsonState.plugin),
-      jsonState.configs.map((configs) => {
+      decodeBase64(jsonState.base64SourceKey),
+      decodeBase64(jsonState.base64PluginKey),
+      jsonState.configIDs.map(configs => {
```
Comment (mohammedsahl) [inline src/state/REPLState.js:81]: Just a heads up these, more likely than not, will change again to S3 links after #32 is closed

File: src/state/detectShare.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  const shareIndicator = "share";
+  return path.includes(shareIndicator);
+}
+
+function extractID() {
+  const path = window.location.pathname;
+  // Attempt to capture everything after the /share/
+  // that's not a URL reserved character as defined here:
+  // https://en.wikipedia.org/wiki/Percent-encoding#Percent-encoding_reserved_characters
+  const re = /\/share\/([^!*'();:@&=+$,/?#[\]]+)/g;
```
Comment (mohammedsahl) [inline src/state/detectShare.js]: Cool regex!

giphy

**Comment by ianjennings** [comment]
Depends on https://github.com/MLH-Fellowship/babel-sandbox-server/pull/12/commits/d6d1e96a1db25c248fa1f989756f9e89346e8556

**Comment by ianjennings** [comment]
Please merge in `master`

File: src/state/REPLState.js [hunk 3]
```diff
@@ -94,13 +78,57 @@ class REPLState {
...
+        headers: {
+          Accept: "application/json",
+          "Content-Type": "application/json",
+        },
+        body: this.Encode(),
+      });
+      const message = await resp.json();
+
+      // https://stackoverflow.com/questions/6941533/get-protocol-domain-and-port-from-url
+      return window.location.href.split('/').slice(0, 3).join('/') + message.url;
```
Comment (ianjennings) [inline src/state/REPLState.js]: `.split('/')` is a really powerful parser when you have repeating characters.

File: src/state/REPLState.js [hunk 4]
```diff
@@ -94,13 +95,54 @@ class REPLState {
...
+    }
+  }
+
+  /**
+   * REPLState.FromID returns a REPLState given a unique identifier.
+   * @param {string} ID
+   * @returns {Promise<REPLState | null>}
+   */
+  static async FromID(ID) {
+    const url = `${REPLState.baseURL}/api/v1/blobs/get-blob/${ID}`;
```
Comment (mohammedsahl) [inline src/state/REPLState.js]: Might need to change this from `/get-blob/${ID}` to `/${ID}` after changes requested in [#12](https://github.com/MLH-Fellowship/babel-sandbox-server/pull/12)
See [route](https://github.com/MLH-Fellowship/babel-sandbox-server/pull/12/files#diff-a9a8d893a8463ac387317e02eec99c96R35)

File: src/state/REPLState.js [hunk 5]
```diff
@@ -40,6 +40,8 @@ function decodeBase64(base64String) {
 }
 
 class REPLState {
+  static baseURL = "http://localhost:1337";
```
Comment (wlawt) [inline src/state/REPLState.js]: Do you think we should set the baseURL in `package.json` by adding a proxy? Ex.:
```
"proxy": "http://localhost:1337"
```

Comment (ianjennings) [reply src/state/REPLState.js]: I didn't know this was a thing! Go for it

Comment (ianjennings) [reply src/state/REPLState.js]: @mohammedsahl is also looking at bundling the client and server

**Review by mohammedsahl — APPROVED** [review]
🔥 LGTM! Awesome job!

**Review by wlawt — APPROVED** [review]

**Pull request merged**